doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,908 @@
+Base (such as address, TX power, channel number) are already stored in RAM.
+
+### Send/Receive
+```c
+    /**
+     * @brief Load packet in the framebuffer of a radio.
+     *
+     * This function shouldn't do any checks, so the packet MUST be valid.
+     * If the radio is still transmitting, it should block until it is safe to
+     * write to the frame buffer again.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] pkt the packet to be sent with valid length
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*prepare)(ieee802154_dev_t *dev, iolist_t *pkt);
+
+    /**
+     * @brief Transmit a preloaded packet.
+     *
+     * @pre the PHY state is @ref IEEE802154_TRX_STATE_TX_ON and the packet
+     *      is already in the framebuffer.
+     *
+     * @post the PHY state is @ref IEEE802154_TRX_STATE_TX_ON.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] mode transmissions mode
+     *
+     * @return 0 on success
+     * @return -ENOTSUP if a transmission mode is not supported
+     * @return -EBUSY if medium is busy (if sending with CCA)
+     * @return negative errno on error
+     */
+    int (*transmit)(ieee802154_dev_t *dev, ieee802154_tx_mode_t mode);
+
+    /**
+     * @brief Get the length of the received packet.
+     *
+     * @pre the radio already received a packet (e.g
+     *      @ref ieee802154_dev_t::cb with @ref IEEE802154_RADIO_RX_DONE).
+     * @pre the device is not sleeping
+     *
+     * @post the frame buffer is still protected against new packet arrivals.
+     *      
+     * @param[in] dev IEEE802.15.4 device descriptor
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*len)(ieee802154_dev_t *dev);
+
+    /**
+     * @brief Read a packet from the internal framebuffer of the radio.
+     *
+     * This function should try to write the received packet into @p buf and
+     * put the radio in a state where it can receive more packets.
+     *
+     * @pre the radio already received a packet (e.g
+     *      @ref ieee802154_dev_t::cb with @ref IEEE802154_RADIO_RX_DONE).
+     * @pre the device is not sleeping
+     *
+     * @post the PHY state is @ref IEEE802154_TRX_STATE_RX_ON and the radio is
+     *       in a state where it can receive more packets, regardless of the
+     *       return value.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[out] buf buffer to write the received packet into. If NULL, the
+     *             packet is not copied.
+     * @param[in] size size of @p buf
+     * @param[in] info information of the received packet (LQI, RSSI). Can be
+     *            NULL if this information is not needed.
+     *
+     * @return number of bytes written in @p buffer (0 if @p buf == NULL)
+     * @return -ENOBUFS if the packet doesn't fit in @p
```

Comment:
Will the packet be discarded here?

Reply:
Frame is the correct term at layer 2. We should try carefully to get those terms correct to avoid confusion. (Sadly, RIOT's doc uses the term packet regardless of the layer at many places. I should `grep` for all usage of `packet` and verify or fix the term.)

Reply:
Yes, the packet is always discarded

Reply:
Actually I forgot something here (thanks for pointing it out @maribu ). All these packets are "Physical Service Data Unit" (PSDU) packets. I will update that

Reply:
I will add a comment that indicates the packet is always discarded

Reply:
PSDU = frame right? Lets decide for one of both. You should add it to the "Acronyms" or "Definitions" section then.

Reply:
I guess PSDU is more explicit. (Frame can have other meanings in other contexts.) Also, PSDU sounds smarter :-)